### DISCUSSION TITLE
Warn about duplicate query parameters

### DISCUSSION THREAD

**Discussion by adius** [Kong/insomnia#3636] [Ideas]
- Insomnia Version: 5.12.4
- Operating System: macOS

## Details

When I accidentally enter the same query parameter twice Insomnia should warn me about it!

**Comment by gschier** [comment]
Where do you think the best place to show a warning would be?

**Comment by adius** [comment]
Maybe put a warning triangle right in front of the key name?

**Comment by gschier** [comment]
Oh ya, that would work. And show the error message in a tooltip on hover? The warning triangle could even replace the drag handle icon so it won't take up any more space.

**Comment by adius** [comment]
Yeah, that's what I thought about. Good idea to replace the icon!

I'm just wondering if this is maybe too subtle. Is there any valid use case where you set a key twice? ^^
If not, maybe we should highlight the whole line yellow and directly write out the warning!

**Comment by gschier** [comment]
Ya, we could change the underline color to yellow as well. I think there are probably use cases where duplicate keys are valid but it's probably rare. If it's too subtle, we can also make use of animation to draw the user's eye (eg. shake icon for 1 second).

**Comment by gschier** [comment]
This could also be applied to the header editor and form editors as well probably

**Comment by adius** [comment]
> underline color to yellow as well

👍

> shake icon for 1 second

Or just pop up the tooltip with the warning for 2 seconds or so and then show it only on hover

**Comment by katanacrimson** [comment]
Honestly, I think this is a bad idea.  My reasoning is as follows:

A query parameter, when repeated, is often used in many languages to handle user input of a List.  For example:

    ?q=a&q=b

would translate to, roughly:

    q = [a, b]

Some languages (such as PHP), or frameworks (such as Rails, iirc) will implement it differently, however.  Arrays are passed in like this: 

    ?q[]=a&q[]=b

Unfortunately, that methodology is completely incompatible with other languages.  For instance, node.js's builtin `querystring`:

    > qs.parse('test=1')
    { test: '1' }
    > qs.parse('test=1&test=2')
    { test: [ '1', '2' ] }
    > qs.parse('test[]=1&test[]=2')
    { 'test[]': [ '1', '2' ] }

So with this feature, you'd actually be nagging at the user to do something that, depending on what their serverside code is written in, they *cannot* always do.

**Comment by adius** [comment]
Well, the case `test[]` could be recognized and ignored, if that's a common pattern.
And for the `qs.parse('test=1&test=2'), { test: [ '1', '2' ] }` case the author can just ignore the warning.
I'd rather have one false positive to ignore, than unlimited cases which break things...

**Comment by adius** [comment]
Or maybe it could even be disabled manually like `right click -> disable warning for this field` or similar

**Comment by gschier** [comment]
I think @damianb has an excellent point. I remember now that I've seen duplicate query parameters used in various web frameworks.

I like the idea of disabling inspections (that's what JetBrains IDEs call them). An inspection being a rule (like duplicate query param names) that generates warnings/errors.

I think this feature would definitely need a way to disable inspections (maybe on the workspace level?). I'm thinking a dropdown menu inside the tooltip that has the option to disable, then also a menu in the workspace settings to toggle the inspections on and off.
